File: Linux/C++/prime4.cpp
#include <iostream>

#define NUM_PRIMES 100

using namespace std;

int main() {
    int number;
    int count;
    int divisor;
    int divisorsquare;
    int primes[NUM_PRIMES];
    int primesquares[NUM_PRIMES];
    int primepointer;
    int remainder;
    bool isprime;

    cout << 2 << endl;
    primes[0] = 2;
    primesquares[0] = 4;
    cout << 3 << endl;
    primes[1] = 3;
    primesquares[1] = 9;
    number = 3;
    for (count = 2; count < NUM_PRIMES; count++) {
        do {
            number = 2 + number;
            primepointer = 1;
            divisor = primes[primepointer];
            divisorsquare = primesquares[primepointer];
            isprime = true;
            while (isprime && divisorsquare <= number) {
                remainder = number % divisor;
                if (remainder == 0) {
                    isprime = false;
                } else {
                    primepointer = 1 + primepointer;
                    divisor = primes[primepointer];
                    divisorsquare = primesquares[primepointer];
                }
            }
            if (isprime) {
                primes[count] = number;
                primesquares[count] = number * number;
                cout << number << endl;
            }
        } while (!isprime);
    }
    return 0;
}

File: Windows/cs/prime4/prime4/prime4.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace prime4 {
    internal class prime4 {
        const uint MAX_PRIMES = 100;

        static void Main() {
            uint count;
            uint divisor;
            ulong divisorsquare;
            int primepointer;
            uint remainder;
            bool isprime;

            uint[] primes = new uint[MAX_PRIMES];
            ulong[] primeSquares = new ulong[MAX_PRIMES];

            primes[0] = 2;
            primeSquares[0] = 4;
            primes[1] = 3;
            primeSquares[1] = 9;
            uint number = 3;

            for (count = 2; count < MAX_PRIMES; count++) {
                do {
                    number += 2;
                    primepointer = 1;
                    divisor = primes[primepointer];
                    divisorsquare = primeSquares[primepointer];
                    isprime = true;

                    while (divisorsquare <= number) {
                        remainder = number % divisor;

                        if (remainder == 0) {
                            isprime = false;
                            break;
                        } else {
                            primepointer++;
                            divisor = primes[primepointer];
                            divisorsquare = primeSquares[primepointer];
                        }
                    }
                    if (isprime) {
                        primes[count] = number;
                        primeSquares[count] = number * number;
                    }
                } while (!isprime);
            }
            for (count = 0; count < MAX_PRIMES; count++) {
                printPrime(count, primes[count]);
            }
        }
        static void printPrime(uint seq, uint number) {
            string outputLine = String.Format("{0,20}: {1,20}", 1 + seq, number);
            Console.WriteLine(outputLine);
        }
    }
}
